runtime/lua/vim/lsp.lua

```diff
@@ -361,6 +364,33 @@ do
     end
   end
 
+  ---@private
+  --
+  -- Adjust debounce time by taking time of last didChange notification into
+  -- consideration. If the last didChange happened more than `debounce` time ago,
+  -- debounce can be skipped and otherwise maybe reduced.
+  --
+  -- This turns the debounce into a kind of client rate limiting
+  local function next_debounce(debounce, state, bufnr)
+    if debounce == 0 then
+      return 0
+    end
+    local ns_to_ms = 0.000001
+    local last_flush = state.last_flush[bufnr]
+    if not last_flush then
+      return debounce
+    end
+    local now = uv.hrtime()
+    local ms_since_last_flush = (now - last_flush) * ns_to_ms
+    local remaining_debounce = (ms_since_last_flush - debounce) * -1
```

Comment:
why not just debounce - ms_since_last_flush?

Reply:
No good reason. I changed it.